### PR TITLE
support case management when editing submissions

### PR DESCRIPTION
**Pull request by czue**
http://manage.dimagi.com/default.asp?123037

easiest read commit-by-commit. i'm not thrilled about the way this got implemented, but it seemed pretty hard to do something that was too much cleaner without significantly changing the internals of how this works.

@dannyroberts is likely most interested/relevant reviewer

most of the actual changes are in https://github.com/dimagi/commcare-hq/commit/23d686073e3092e36968176ead5b44580b3286d5

(wait for tests before merge, obviously)

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/case/xform.py [hunk 1]
```diff
@@ -7,13 +7,14 @@
 from casexml.apps.case.signals import cases_received, case_post_save
 from corehq.toggles import LOOSE_SYNC_TOKEN_VALIDATION
 from casexml.apps.case.util import iter_cases
-from couchforms.models import XFormInstance
+from couchforms.models import XFormInstance, XFormDeprecated
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/xform.py]: F401 'XFormDeprecated' imported but unused

File: corehq/ex-submodules/couchforms/util.py [hunk 2]
```diff
@@ -414,7 +422,10 @@ def process(xform):
             return self.get_exception_response(e.error_log), None, []
         else:
             from casexml.apps.case.models import CommCareCase
-            from casexml.apps.case.xform import get_and_check_xform_domain, CaseDbCache, process_cases_with_casedb
+            from casexml.apps.case.xform import (
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/util.py]: F401 'process_cases_with_casedb' imported but unused

**Comment by czue** [comment]
:white_check_mark:

**Comment by dannyroberts** [comment]
Oooh, nice! I remember this guy

File: corehq/ex-submodules/casexml/apps/case/models.py [hunk 3]
```diff
@@ -583,7 +585,13 @@ def _safe_replace_and_force_to_string(me, attr, val):
     
     def update_from_case_update(self, case_update, xformdoc, other_forms=None):
         other_forms = other_forms or {}
-        if case_update.has_referrals():
+        if is_deprecation(xformdoc):
+            # remove the form from actions and xform_ids (based on orig_id)
+            # and rebuild
+            self.xform_ids = [id for id in self.xform_ids if id != xformdoc.orig_id]
+            self.actions = [a for a in self.actions if a.xform_id != xformdoc.orig_id]
```
Comment (dannyroberts) [inline corehq/ex-submodules/casexml/apps/case/models.py]: I think there's a design decision here which is to undo the original case action by removing and rebuilding, and then applying the new case action on top of the whole thing, even if there are intervening case actions between the original form submission and the second, deprecating, form submission. Am I reading that right? That seems reasonable—maybe also leave a comment explaining that decision (as opposed to other other plausible decision of swapping in the new case action for the old one in the same position in the list).

**Comment by dannyroberts** [comment]
My review is complete. This looks really good, just left a few small comments. Actually pretty impressed how cleanly you were able to handle this, and the cleanup and test-driven approach both seemed super positive

**Comment by czue** [comment]
responded to feedback

File: corehq/ex-submodules/casexml/apps/case/models.py (hunk 3 above)
Comment (czue) [inline corehq/ex-submodules/casexml/apps/case/models.py]: it wasn't a completely explicit design decision so much as it looked a lot easier to implement. can add a comment if you think it's useful though.

Comment (dannyroberts) [inline corehq/ex-submodules/casexml/apps/case/models.py]: Yeah I think worth it. Then I think this is ready to deploy!

Comment (czue) [inline corehq/ex-submodules/casexml/apps/case/models.py]: is this approximately what you had in mind? https://github.com/dimagi/commcare-hq/commit/9133144d319be2d7b6249d0f7ef96d741c1fb755

after writing that i decided I wanted to write a test to make sure it's valid so feel free to hold off on merge for a bit while I work on that.

File: corehq/ex-submodules/couchforms/tests/test_edits.py [hunk 4]
```diff
@@ -123,3 +130,124 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+            case_type='person',
+            owner_id=owner_id,
+            version=V2,
+        ).as_string(format_datetime=json_format_datetime)
+        create_form_id = submit_case_blocks(case_block, domain=self.domain)
+
+        # validate that worked
+        case = CommCareCase.get(case_id)
+        self.assertEqual([create_form_id], case.xform_ids)
+        self.assertEqual([create_form_id],  [a.xform_id for a in case.actions])
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/tests/test_edits.py]: E241 multiple spaces after ','

**Comment by czue** [comment]
@dannyroberts if you agree with https://github.com/dimagi/commcare-hq/pull/6040#commitcomment-10399110 then don't merge this. I'll look into fixing it up.

**Comment by dannyroberts** [comment]
@czue it's hard to say, but when I originally saw it in this PR and interpreted it that way, it didn't strike me as wrong. It struck me as one of the two plausible ways to handle it, and it's hard to say which one's better. This was seems the most "consistent" in that we sort everything else by submit date and don't want to, like, spoof the submit date on deprecating forms just to get around that.

**Comment by czue** [comment]
actually I take it back. I think the case processing logic is ok and the right thing for HQ to do is introduce a concept of "server_edit_time" on the form and preserve the order using the original submission timestamp. what do you think?

**Comment by czue** [comment]
ha, i just proposed the thing you said we shouldn't do, but i actually think it's right from a data perspective. let's chat in person if you disagree

File: corehq/ex-submodules/casexml/apps/case/xml/parser.py [hunk 5]
```diff
@@ -3,11 +3,23 @@
 objects from things from xforms.
 """
 import os
+import datetime
 
 from casexml.apps.case import const
+from casexml.apps.case.models import CommCareCaseAction
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/xml/parser.py]: F401 'CommCareCaseAction' imported but unused

File: corehq/ex-submodules/casexml/apps/case/xml/parser.py [hunk 6]
```diff
@@ -55,9 +67,7 @@ def __init__(self, block, type=None, name=None, external_id=None,
         self.attachments = attachments or {}
     
     def get_known_properties(self):
-        prop_list = ["type", "name", "external_id", "user_id",
-                     "owner_id", "opened_on"]
-        return dict((p, getattr(self, p)) for p in prop_list \
+        return dict((p, getattr(self, p)) for p in KNOWN_PROPERTIES.keys() \
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/xml/parser.py]: E502 the backslash is redundant between brackets

**Comment by czue** [comment]
@dannyroberts I think this is good to go pending tests. I ended up rewriting a lot about how case processing works in the process and I think it's quite a bit nicer. tried to break up by commit, but let me know if you want to pair review it.

**Comment by millerdev** [comment]
Test fail, not sure if this is real
```
======================================================================
FAIL: testForA_B (pact.tests.dot_ordering.dotsOrderingTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "custom/_legacy/pact/tests/dot_ordering.py", line 114, in testForA_B
    self.testFormA()
  File "custom/_legacy/pact/tests/dot_ordering.py", line 96, in testFormA
    self._submitAndVerifyBundle(bundle)
  File "custom/_legacy/pact/tests/dot_ordering.py", line 145, in _submitAndVerifyBundle
    self._verify_dot_cells(nonart_submissions, art_submissions, examine_day)
  File "custom/_legacy/pact/tests/dot_ordering.py", line 149, in _verify_dot_cells
    self.assertEqual(examine_day[DOT_NONART_IDX][day_dose_idx][0:2], nonart_submissions[day_dose_idx][0:2])
AssertionError: Lists differ: ['unchecked', 'pillbox'] != ['full', 'pillbox']

**Comment by czue** [comment]
@millerdev was a classing pact evening fail. :white_check_mark: now

**Comment by dannyroberts** [comment]
@czue going to review the new changes today

**Comment by czue** [comment]
woot

**Pull request merged**